ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -43,27 +48,45 @@ private[platform] case class StateCache[K, V](cache: Cache[K, V], registerUpdate
     * this method updates the cache only if the to-be-inserted tuple is the most recent
     * (i.e. it has `validAt` highest amongst the competing updates).
     *
+    * NOTE: Calls to `put` evolve the cache index. Callers should ensure strictly increasing validAt evolution.
+    *
     * @param key the key at which to update the cache
     * @param validAt ordering discriminator for pending updates for the same key
     * @param value the value to insert
     */
-  def put(key: K, validAt: Long, value: V): Unit = Timed.value(
-    registerUpdateTimer, {
-      pendingUpdates.synchronized {
-        val competingLatestForKey =
-          pendingUpdates
-            .get(key)
-            .map { pendingUpdate =>
-              val oldLatestValidAt = pendingUpdate.latestValidAt
-              pendingUpdate.latestValidAt = Math.max(validAt, pendingUpdate.latestValidAt)
-              oldLatestValidAt
-            }
-            .getOrElse(Long.MinValue)
+  def put(key: K, validAt: Long, value: V)(implicit loggingContext: LoggingContext): Unit =
+    Timed.value(
+      registerUpdateTimer, {
+        pendingUpdates.synchronized {
+          // Synchronous updates should generally increase the _cacheIndex strictly monotonic.
+          // However, on transient DB errors, the restart of the update flow can lead to
+          // re-update at the current _cacheIndex.
+          if (validAt >= _cacheIndex) {
+            val competingLatestForKey =
+              pendingUpdates
+                .get(key)
+                .map { pendingUpdate =>
+                  val oldLatestValidAt = pendingUpdate.latestValidAt
+                  pendingUpdate.latestValidAt = Math.max(validAt, pendingUpdate.latestValidAt)
+                  oldLatestValidAt
+                }
+                .getOrElse(Long.MinValue)
 
-        if (competingLatestForKey < validAt) cache.put(key, value) else ()
-      }
-    },
-  )
+            if (competingLatestForKey < validAt) putInternal(key, value, validAt) else ()
+
+            _cacheIndex = validAt
+          } else {
+            // Updating the cache synchronously before the _cacheIndex is an error scenario.
+            // Abort pending updates and invalidate the cache entry.
+            pendingUpdates -= key
+            cache.invalidate(key)
+            logger.error(
```

Comment:
I am not sure about the intent here. The condition is also strange: in case of stream is coming in a steady fashion, then also == should be grave error. If we have resubscription see my comment on line 62. I also have mixed feelings about the action here: an error, but letting the stream go.
Can it be that this is just a warning, and we need only > at line 64?
Can it be that we want to throw errors and letting the stream die if for one continuous stream we see non-monotonicity?
Can it be that we want to be more explicit about the stream-resubscriptions?

Reply:
You are right. There is room for simplification. Discussed offline and concluded that synchronous updates coming before the cache index can be ignored. Updated